### ISSUE TITLE
False Positive: nonminimal_bool suggesting a function which doesn't exist.

### ISSUE THREAD

**Issue by Hoverbear** [rust-lang-nursery/rust-clippy#2638]
Hi!

We're upgrading clippy as part of https://github.com/pingcap/tikv/pull/2922 and I'm seeing what I think are some false positives.

We have a type that implements `is_ok()`, but not something such as `is_err()`:

```rust
pub enum Res<T> {
    Ok(T),
    Truncated(T),
    Overflow(T),
}

impl<T> Res<T> {
    pub fn map<U, F: FnOnce(T) -> U>(self, f: F) -> Res<U> {
        match self {
            Res::Ok(t) => Res::Ok(f(t)),
            Res::Truncated(t) => Res::Truncated(f(t)),
            Res::Overflow(t) => Res::Overflow(f(t)),
        }
    }

    pub fn unwrap(self) -> T {
        match self {
            Res::Ok(t) | Res::Truncated(t) | Res::Overflow(t) => t,
        }
    }

    pub fn is_ok(&self) -> bool {
        match *self {
            Res::Ok(_) => true,
            _ => false,
        }
    }

    pub fn is_overflow(&self) -> bool {
        match *self {
            Res::Overflow(_) => true,
            _ => false,
        }
    }

    pub fn is_truncated(&self) -> bool {
        match *self {
            Res::Truncated(_) => true,
            _ => false,
        }
    }
}
```

Where we use `if !res.is_ok() {}` we get clippy lints like this:

```rust
warning: this boolean expression can be simplified
    --> src/coprocessor/codec/mysql/decimal.rs:1922:12
     |
1922 |         if !res.is_ok() {
     |            ^^^^^^^^^^^^ help: try: `res.is_err()`
     |
     = help: for further information visit https://rust-lang-nursery.github.io/rust-clippy/v0.0.192/index.html#nonminimal_bool
```

Clippy is suggesting a function which doesn't exist, and this lint seems to be misfiring. Changing the function name from `is_ok()` fixes the problem.

Example: https://play.rust-lang.org/?gist=8d887a30807a1d21a471be0998d52995&version=stable

**Comment by oli-obk** [comment]
ha, nice catch. Needs a type check for `Result`

**Comment by phansch** [comment]
Going to take a look at this as I only worked with macro/attribute related stuff so far.

**Comment by Hoverbear** [comment]
Thanks for tackling this! :) Really appreciate it!
